src/arcanum.rs

```diff
@@ -0,0 +1,93 @@
+use crate::util::Result;
+use html2md::parse_html;
+use reqwest::Url;
+use serde::Deserialize;
+use serenity::builder::CreateEmbed;
+
+pub struct Arcanum {
+    base_url: String,
+    icon: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct Entry {
+    id: u64,
+    event: u64,
+    event_name: String,
+    event_date: String,
+    date: String,
+    paraphrased: bool,
+    modified_date: String,
+    tags: Vec<String>,
+    lines: Vec<Line>,
+    note: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct PaginatedEntries {
+    pub count: u64,
+    next: Option<String>,
+    previous: Option<String>,
+    pub results: Vec<Entry>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct Line {
+    speaker: String,
+    text: String,
+}
+
+impl Arcanum {
+    async fn raw_request(&self, endpoint: &str, params: &Vec<(&str, &str)>) -> Result<String> {
+        let base_url = format!("{}{}", self.base_url, endpoint);
+        let url = Url::parse_with_params(&base_url, params).expect("Invalid URL or parameters");
+
+        println!("URL: {url}");
+
+        let response = reqwest::get(url).await?.text().await?;
+        Ok(response)
+    }
+
+    pub fn embed_entry(&self, entry: &Entry) -> CreateEmbed {
+        // TODO: this should check for size, etc
+        CreateEmbed::new()
+            .title(&entry.event_name)
+            .color((0, 99, 91))
+            .url(format!(
+                "{}/events/{}/#e{}",
+                self.base_url, entry.event, entry.id
+            ))
+            .thumbnail(&self.icon)
+            .fields(
+                entry
+                    .lines
+                    .iter()
+                    .map(|Line { speaker, text }| (parse_html(speaker), parse_html(text), false)),
+            )
+    }
+
+    pub async fn random_entry(&self) -> Result<Entry> {
+        Ok(serde_json::from_str(
+            &self.raw_request("/api/random_entry", &vec![]).await?,
+        )?)
```

Comment:
```suggestion
        serde_json::from_str(
            &self.raw_request("/api/random_entry", &vec![]).await?,
        ).map_err(Into::into)
```
I believe these are equivalent, and I think the `map_err` is a little cleaner. Up to you on this one, it's a minor style thing. If you do change it, same for `search_entries`.